picoquic/frames.c

```diff
@@ -1975,19 +2135,26 @@ int picoquic_decode_frames(picoquic_cnx_t* cnx, uint8_t* bytes,
         size_t consumed = 0;
 
         if (first_byte >= picoquic_frame_type_stream_range_min && first_byte <= picoquic_frame_type_stream_range_max) {
-            ret = picoquic_decode_stream_frame(cnx, bytes + byte_index,
-                bytes_max - byte_index, restricted, &consumed, current_time);
-            cnx->ack_needed = 1;
-            byte_index += consumed;
+            if (epoch != 1 && epoch != 3) {
+                DBG_PRINTF("Data frame (0x%x), when only TLS stream is expected", first_byte);
+                ret = picoquic_connection_error(cnx, PICOQUIC_TRANSPORT_PROTOCOL_VIOLATION);
+            } else {
+                ret = picoquic_decode_stream_frame(cnx, bytes + byte_index,
+                    bytes_max - byte_index, &consumed, current_time);
+                cnx->ack_needed = 1;
+                byte_index += consumed;
+            }
         } else if (first_byte == picoquic_frame_type_ack) {
             ret = picoquic_decode_ack_frame(cnx, bytes + byte_index,
-                bytes_max - byte_index, &consumed, current_time, restricted);
+                bytes_max - byte_index, &consumed, current_time, epoch);
             byte_index += consumed;
         } else {
-            if (restricted && first_byte != picoquic_frame_type_padding 
+            if (epoch != 1 && epoch != 3 && first_byte != picoquic_frame_type_padding 
```

Comment:
Would an enum be better than magic numbers for epoch?

Reply:
Yes, probably, more mnemonic. I was lazy.